docs/proposals/crt_management_proposal.md

```diff
@@ -0,0 +1,113 @@
+# Container Runtime Management
+
+## Description
+origin and openshift-ansible support multiple container runtimes.  This proposal
+is related to refactoring how we handle those runtimes in openshift-ansible.
+
+### Problems addressed
+We currently don't install docker during the install at a point early enough to
+not fail health checks, and we don't have a good story around when/how to do it.
+This is complicated by logic around containerized and non-containerized installs.
+
+A web of dependencies can cause changes to docker that are unintended and has
+resulted in a series of work-around such as 'skip_docker' boolean.
+
+We don't handle docker storage because it's BYO.  By moving docker to a prerequisite
+play, we can tackle storage up front and never have to touch it again.
+
+container_runtime logic is currently spread across 3 roles: docker, openshift_docker,
+and openshift_docker_facts.  The name 'docker' does not accurately portray what
+the role(s) do.
+
+## Rationale
+* Refactor docker (and related meta/fact roles) into 'container_runtime' role.
+* Strip all meta-depends on container runtime out of other roles and plays.
+* Create a 'prerequisites.yml' entry point that will setup various items
+such as container storage and container runtime before executing installation.
+* All other roles and plays should merely consume container runtime, should not
+configure, restart, or change the container runtime as much as feasible.
+
+## Design
+
+The container_runtime role should be comprised of 3 'pseudo-roles' which will be
+consumed using include_role; each component area should be enabled/disabled with
+a boolean value, defaulting to true.
+
+I call them 'pseudo-roles' because they are more or less independent functional
+areas that may share some variables and act on closely related components.  This
+is an effort to reuse as much code as possible, limit role-bloat (we already have
+an abundance of roles), and make things as modular as possible.
+
+```yaml
+# prerequisites.yml
+- include: std_include.yml
+- include: container_runtime_setup.yml
+...
+# container_runtime_setup.yml
+- hosts: "{{ openshift_runtime_manage_hosts | default('oo_nodes_to_config') }}"
+  tasks:
+    - include_role:
+        name: container_runtime
+        tasks_from: install.yml
```

Comment:
I'm not understanding the reasoning of calling this role three times to run different pieces.  It seems that it could be called once and the logic in the role would perform the steps required.  Do these parts need to be called independently in other areas?

Reply:
Just more flexibility for the user.  I can imagine a case where someone might wish to do some of these parts and not others (such as configuring storage).

We could also put similar logic in the main.yml of the role itself.

We have mentioned doing 'role decomposition' in other areas, this is a continuation of that pattern.  I believe skipping an include_role at the play level will prevent a lot of skip tasks vs putting the login in main.yml of the role, but I'm not 100% sure on that.

Reply:
I like the ability to call each section separately.  @mtnbikenc, is there some inefficiency when making multiple calls that outweighs the decomposition?  If they are similar in speed I'd prefer the separate calls.

Otoh, Since we are in prerequisites and running a single time then I could see the skips not being a big deal.  Maybe have main call each separate piece and continue to keep the underlying pieces broken up in case users want to call them separately.

I'm fine either way here.